### PR TITLE
speed up url.c++ scheme and port detection

### PR DESCRIPTION
**Pull request by anonrig**
While implementing Ada, we found out that checking if URL is special or not, and getting the default port of a scheme had extremely high cost. Let's apply the same optimizations we did on Ada, on workerd.

### CODE REVIEW

**Review by jasnell — APPROVED** [review]
LGTM. This is the version that is used in the old original URL implementation that is no longer on by default but there are still lots of workers using this so I'm sure they'll appreciate the slight performance bump :-)

File: src/workerd/api/url.c++ [hunk 1]
```diff
@@ -17,26 +17,38 @@ namespace {
...
+  return (target[0] == scheme[0]) && (target.slice(1) == scheme.slice(1));
+}
+
+// Taken from Ada URL library.
+// Ref: https://github.com/ada-url/ada/blob/b431670699cf4f3ebb2e2c394c23a89850bb6f3f/include/ada/scheme-inl.h#L57
+kj::Maybe<kj::StringPtr> defaultPortForScheme(kj::StringPtr scheme) noexcept {
+  if (scheme.size() == 0) {
+    return kj::none;
+  }
+  int hash_value = (2 * scheme.size() + (unsigned)(scheme[0])) & 7;
```
Comment (npaun) [inline src/workerd/api/url.c++]:
```suggestion
  int hash_value = (2 * scheme.size() + static_cast<unsigned>(scheme[0])) & 7;
```

**Review by npaun — APPROVED** [review]

File: src/workerd/api/url.c++ [hunk 2]
```diff
@@ -17,26 +17,38 @@ namespace {
...
-    {"ws", "80"},
-    {"wss", "443"},
-  };
-  auto port = defaultPorts.find(scheme);
-  if (port != defaultPorts.end()) {
-    return port->second;
+constexpr kj::StringPtr is_special_list[] = {
+  "http"_kj, " "_kj, "https"_kj, "ws"_kj, "ftp"_kj, "wss"_kj, "file"_kj, " "_kj};
+constexpr kj::StringPtr special_ports[] = {
+  "80"_kj, ""_kj, "443"_kj, "80"_kj, "21"_kj, "443"_kj, ""_kj, ""_kj};
```
Comment (jasnell) [inline src/workerd/api/url.c++:23]: nit: a comment here explaining to keep these two lists in sync would be helpful.

File: src/workerd/api/url.c++ [hunk 3]
```diff
@@ -17,26 +17,38 @@ namespace {
...
+
+// Taken from Ada URL library.
+// Ref: https://github.com/ada-url/ada/blob/b431670699cf4f3ebb2e2c394c23a89850bb6f3f/include/ada/scheme-inl.h#L49
+bool isSpecialScheme(kj::StringPtr scheme) noexcept {
+  if (scheme.size() == 0) {
+    return false;
+  }
+  int hash_value = (2 * scheme.size() + static_cast<unsigned>(scheme[0])) & 7;
```
Comment (jasnell) [inline src/workerd/api/url.c++:31]: nit: A comment explaining this hash function would likely be worthwhile.

**Review by jasnell — APPROVED** [review]

**Pull request merged**